aksolve/client/aksolve/conjugate_gradients.py

```diff
@@ -0,0 +1,218 @@
+        Ap = matvec(p)
+        alfa = numer / inner(p, Ap)
+
+        x = x + alfa * p
+
+        if niter % irestart == 0:
+            r = b - matvec(x)
+        else:
+            r = r - alfa * Ap
+
+        rnorm = norm(r)
+        s = precon(r)
+        numer, denom = inner(r, s), numer
+        p = s + (numer / denom) * p
+        niter += 1
+
+        if verbose: print(f'{niter:>10} {rnorm:>10.2e}')
+
+    if niter >= max_iter : istop = 4
+    if rnorm < tol       : istop = 3
+    if rnorm < EPS       : istop = 2
+
+    if verbose:
+        print('\nExit Conjugate Gradient Solver.\n')
+        print(f'Status: {stop_status[istop]}')
+        print(f' niter = {niter}\n'
+              f' rnorm = {rnorm:0.3e}\n')
+
+    return (x, istop, niter, rnorm)
+
+
+if __name__ == '__main__':
+    from aksolve.util import matvec_from_coo
+
+    print('####################')
+    print('# PERFORMING TESTS #')
+    print('####################')
+
+    host = input("enter arkouda hostname: ")
+    ak.connect(host)
+
+    print('==== Testing Conjugate Gradients ====')
+    # A = [[4, 1],
+    #      [1, 3]]
+    R = ak.array([0, 0, 1, 1])
+    C = ak.array([0, 1, 0, 1])
+    V = ak.array([4, 1, 1, 3])
+
+    x = ak.array([1/11, 7/11])  # true answer
+    b = ak.array([1, 2])        # RHS
+
+    matvec = matvec_from_coo(R, C, V)
+
+    print('\n--> RHS == 0\n')
+    ans = cg(matvec, ak.zeros(2), verbose=True)
+    assert ans[1] == 0
+
+    print('\n--> Ax == b\n')
+    ans = cg(matvec, b, x)
+    assert ans[1] == 1 and ak.all(ans[0] == x)
+
+    print('\n--> Typical Solving\n')
+    ans = cg(
+        matvec,
+        b,
+        x_start=ak.array([2, 1]),
+        max_iter=5,
+        verbose=True)
+    assert ak.all(ak.abs(ans[0] - x) < 1.0e-4)
+    assert ans[3] < 1.0e-6
+    assert ans[2] == 2
+    assert ans[1] == 3
+
+    print('\n--> Solve to Machine Precision\n')
+    ans = cg(
+        matvec,
+        b,
+        x_start=ak.array([2, 1]),
+        tol=EPS,
+        max_iter=5,
+        verbose=True)
+    assert ak.all(ak.abs(ans[0] - x) < 1.0e-4)
+    assert ans[3] < EPS
+    assert ans[2] == 3
+    assert ans[1] == 2
+
+    print('\n--> Solve with Diagonal Preconditioning\n')
+    P = lambda x: ak.array([1/4, 1/3]) * x
+    ans = cg(matvec, b, precon=P, verbose=True)
+    assert ak.all(ak.abs(ans[0] - x) < 1.0e-4)
+    assert ans[3] < EPS
+    assert ans[2] == 2
+    assert ans[1] == 2
+
+    print('\n--> Solve: Large Diagonal matrix with Diagonal Preconditioning\n')
+    D = ak.randint(1, 1000, 2 ** 32, dtype='float64')
+    matvec = lambda x: x * D
+    x = ak.randint(0, 1, 2 ** 32, dtype='float64')
+    P = lambda x: x * (1.0 / D)
+    b = matvec(x)
+    ans = cg(matvec, b, precon=P, verbose=True)
+    assert ak.all(ak.abs(ans[0] - x) < 1.0e-4)
+    assert ans[3] < 1.0e-6
+    assert ans[2] == 1
+    assert ans[1] == 3
+
+    print('##############')
+    print('# YOU PASSED #')
+    print('##############')
```

Comment:
This needs to be moved into a `test` directory (same level as client). Configure to use pytest.